### ISSUE TITLE
Orbital Survey bypasses hard drives

### ISSUE THREAD

**Issue by SirMortimer** [Kerbalism/Kerbalism#399]
<a href="https://github.com/UsernamesAreAlwaysTaken"><img src="https://avatars3.githubusercontent.com/u/48958828?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [UsernamesAreAlwaysTaken](https://github.com/UsernamesAreAlwaysTaken)**
_Tuesday May 07, 2019 at 16:41 GMT_
_Originally opened as https://github.com/SirMortimer/Kerbalism/issues/76_

----

![asdasdasdasd](https://user-images.githubusercontent.com/48958828/57317028-0f6ff180-70ef-11e9-8c58-75a4dde4f194.gif)

I believe the image is self-explanatory.

The issue that arises from this: no background tranmission is possible. This becomes major when you factor in the antenna bandwidth nerfs. For kerbin, with a RA-100, you'd probably need an orbit or 2, while having the vessel active, to transmit the data back.

Now think about eelo. (probably timewarp for a few months with vessel active to transmit data. If  connection is lost, afaik you have to restart from scratch. Fun times.

**Comment by SirMortimer** [comment]
<a href="https://github.com/gotmachine"><img src="https://avatars1.githubusercontent.com/u/24925209?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [gotmachine](https://github.com/gotmachine)**
_Tuesday May 07, 2019 at 19:24 GMT_

----

Not the case, the stock module completely bypass our transmission rates, it uses the stock rates and is unaffected by Kerbalism.
Still, we could have add some support for the stock scanner.
It should possible to make a derived class from the ModuleOrbitalSurveyor and add lightweight support since all the methods are virtual.

Specifically we should :
- override OnAwake with the following : 
`perfSurv = base.Events["PerformSurvey"];` 
- override OnDestroy with an empty method
- override sendDataToComms() and make it create (instantly) a file with a size of :
 `vessel.mainBody.Radius * 0.001 / ScanTime;`
  with the subject_id :
`"survey@" + vessel.mainBody.bodyName`
- add the following to a static method called in Science.Update when the file has been fully transmitted : 
```
post message : Localizer.Format("#autoLOC_259361", body.displayName);
ResourceMap.Instance.UnlockPlanet(body.flightGlobalsIndex);
```

**Comment by SirMortimer** [comment]
<a href="https://github.com/UsernamesAreAlwaysTaken"><img src="https://avatars3.githubusercontent.com/u/48958828?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [UsernamesAreAlwaysTaken](https://github.com/UsernamesAreAlwaysTaken)**
_Saturday May 11, 2019 at 13:27 GMT_

----

override sendDataToComms() and make it create (instantly) a file with a size of :
vessel.mainBody.Radius * 0.001 / ScanTime;

This would work, but IMO make the 0.001 multiplier patchable for balance reasons. Some random orbital experiments would take up more drive capacity than a survey, which would make little sense. The survey is probably by far the most useful bit of info, so it should have a pretty high data scale. Instead of me asking you to change the code every time, let me patch it through MM, because i assume it'll take a few attempts till we get it right.

Another point, I'm not sure how it'll interract with orbital survey plus, as ScanTime is probably messed with (gradually generate overlays based on sat position). Remove scan time, as stock is 5s for all afaik, so it's just a flat x5. If needed, it can be factored in the multiplier above.

Question: As it'll be forced through the drives now, would our data tranmission rates kick in? Or would it have it's own transmission mechanics?

**Comment by SirMortimer** [comment]
<a href="https://github.com/SirMortimer"><img src="https://avatars2.githubusercontent.com/u/43166475?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [SirMortimer](https://github.com/SirMortimer)**
_Thursday May 16, 2019 at 07:52 GMT_

----

Anything that goes through our drives is subjected to our transmission rates.

I'd suggest something that goes a bit further:
- Create our own part module for orbital surveys, as @gotmachine described above
- Add to that a background update method that just generates data at a configured rate (that could later be made depending on the type of orbit)

For patching/balancing purposes, the total file size and data generation rates should be configurable with our extended scanner part module. Also, the file size should be depending on the size of the body.

Unlock the overlay map only when the file has been fully-ish (maybe min. 90% received?) transmitted, somewhere in Science.Update.

**Comment by SirMortimer** [comment]
<a href="https://github.com/UsernamesAreAlwaysTaken"><img src="https://avatars3.githubusercontent.com/u/48958828?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [UsernamesAreAlwaysTaken](https://github.com/UsernamesAreAlwaysTaken)**
_Thursday May 16, 2019 at 21:10 GMT_

----

Works for me

If possible, I'd like to request splitting up the resources one by one. As in, let me make a configure module, to select a resource the player should scan for, and once transmitted, the overlay should only be generated for that specific resource. All the others should still be locked if not specifically scanned for.

I'm not sure, does Nitrogen, Oxygen, whatever other resources kerbalism uses have their own overlays? The filters/harvesters somehow account for concentrations, therefore the data is present somewhere i assume.

**Comment by gotmachine** [comment]
A few remarks :
- There is no easy way to unlock the resources one by one with the stock scanner. That would require a much more complex system and hacking of the stock GUI.
- OrbitalSurveyPlus implement its own partmodules and its own overlays, support for it would be completely different (much more like scansat support). I don't think it's possible to unlock the ressources individually with it either.

**Comment by gotmachine** [comment]
fixed in a23ab4c82ce4179a9a2233aefdc20d23c03d3fcd
